lib/client/keystore.go

```diff
@@ -203,493 +222,355 @@ func (fs *FSLocalKeyStore) AddKey(host, username string, key *Key) error {
-		kubeDir := filepath.Join(dirPath, idx.username+kubeDirSuffix, o.teleportClusterName)
-		kubeFiles, err := ioutil.ReadDir(kubeDir)
-		if err != nil && !os.IsNotExist(err) {
-			return trace.Wrap(err)
-		}
-		if key.KubeTLSCerts == nil {
-			key.KubeTLSCerts = make(map[string][]byte)
+	if info.IsDir() {
+		certDataMap := map[string][]byte{}
+		certFiles, err := ioutil.ReadDir(certPath)
+		if err != nil {
+			return nil, trace.ConvertSystemError(err)
 		}
-		for _, fi := range kubeFiles {
-			data, err := ioutil.ReadFile(filepath.Join(kubeDir, fi.Name()))
+		for _, certFile := range certFiles {
+			data, err := ioutil.ReadFile(filepath.Join(certPath, certFile.Name()))
 			if err != nil {
-				return trace.Wrap(err)
+				return nil, trace.ConvertSystemError(err)
 			}
-			kubeCluster := strings.TrimSuffix(filepath.Base(fi.Name()), fileExtTLSCert)
-			key.KubeTLSCerts[kubeCluster] = data
+			name := strings.TrimSuffix(certFile.Name(), fileExtTLSCert)
+			certDataMap[name] = data
 		}
-	case *MemLocalKeyStore:
-		idx.clusterName = o.teleportClusterName
-		stored, ok := s.inMem[idx]
-		if !ok {
-			return trace.NotFound("key for %v not found", idx)
-		}
-		key.KubeTLSCerts = stored.KubeTLSCerts
-	default:
-		return trace.BadParameter("unexpected key store type %T", store)
+		return certDataMap, nil
 	}
-	if key.ClusterName == "" {
-		key.ClusterName = o.teleportClusterName
+
+	certData, err := ioutil.ReadFile(certPath)
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
 	}
-	return nil
+	return certData, nil
+}
+
+// CertOption is an additional step to run when loading/deleting user certificates.
+type CertOption interface {
+	// relativeCertPath returns a path to the cert (or to a dir holding the certs)
+	// relative to the session key dir. For use with FSLocalKeyStore.
+	relativeCertPath(idx KeyIndex) string
+	// updateKey adds the cert data to the key and performs related checks.
+	updateKey(key *Key, certData interface{}) error
+	// deleteFromKey deletes the cert data from the key.
+	deleteFromKey(key *Key)
+}
+
+// WithAllCerts lists all known CertOptions.
+var WithAllCerts = []CertOption{WithSSHCerts{}, WithKubeCerts{}, WithDBCerts{}, WithAppCerts{}}
+
+// WithSSHCerts is a CertOption for handling SSH certificates.
+type WithSSHCerts struct{}
+
+func (o WithSSHCerts) relativeCertPath(idx KeyIndex) string {
+	components := []string{idx.ProxyHost, idx.Username + sshDirSuffix}
+	if idx.ClusterName != "" {
+		components = append(components, idx.ClusterName+constants.FileExtSSHCert)
+	}
+	return filepath.Join(components...)
 }
 
-func (o withKubeCerts) deleteKey(store LocalKeyStore, idx keyIndex) error {
-	switch s := store.(type) {
-	case *FSLocalKeyStore:
-		dirPath := s.dirFor(idx.proxyHost)
-		kubeCertsDir := filepath.Join(dirPath, idx.username+kubeDirSuffix, o.teleportClusterName)
-		if err := os.RemoveAll(kubeCertsDir); err != nil {
+func (o WithSSHCerts) updateKey(key *Key, certData interface{}) error {
```

Comment:
Let's just pass []byte, practice shows that returning/accepting interface{} is almost never a good idea.